### PR TITLE
MAINT: Cleanup more NPY_PY3K

### PR DESCRIPTION
**Pull request by sethtroisi**
These were the not-obviously-trivial cases for removing NPY_PY3K I've tried to annotate what and why.

### CODE REVIEW

File: numpy/core/src/multiarray/alloc.c [hunk 1]
```diff
@@ -48,11 +48,7 @@ static cache_bucket datacache[NBUCKETS];
 static cache_bucket dimcache[NBUCKETS_DIM];
 
 /* as the cache is managed in global variables verify the GIL is held */
-#if defined(NPY_PY3K)
 #define NPY_CHECK_GIL_HELD() PyGILState_Check()
```
Comment (sethtroisi) [inline numpy/core/src/multiarray/alloc.c]: This is only used in two places inside this file ([see here](https://github.com/numpy/numpy/blob/950dd4e15ea0976bb671148440e036c3ae2dc11d/numpy/core/src/multiarray/alloc.c#L70)). I'm happy to inline it if people want.

File: numpy/core/src/multiarray/arrayobject.c [hunk 2]
```diff
@@ -1004,22 +1004,18 @@ _strings_richcompare(PyArrayObject *self, PyArrayObject *other, int cmp_op,
...
     if (PyArray_TYPE(self) != PyArray_DESCR(other)->type_num) {
-#if defined(NPY_PY3K)
         /*
          * Comparison between Bytes and Unicode is not defined in Py3K;
          * we follow.
          */
         Py_INCREF(Py_NotImplemented);
         return Py_NotImplemented;
-#else
-        cast = 1;
```
Comment (sethtroisi) [inline numpy/core/src/multiarray/arrayobject.c]: this involved cleaning up cast so I pulled it out of #15281

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 3]
```diff
@@ -1530,7 +1530,7 @@ OBJECT_to_@TOTYPE@(void *input, void *output, npy_intp n,
  * #convert = 1*18, 0*3, 1*2,
  *            1*18, 0*3, 1*2,
  *            0*23#
- * #convstr = (Int*9, Long*2, Float*4, Complex*3, Tuple*3, Long*2)*3#
+ * #convstr = (Long*9, Long*2, Float*4, Complex*3, Tuple*3, Long*2)*3#
```
Comment (sethtroisi) [inline numpy/core/src/multiarray/arraytypes.c.src]: #define PyInt_type PyLong_type changes these Int's to Long's

took me a second to understand this logic (also to see if I could remove IS_STRING)

File: numpy/core/src/multiarray/conversion_utils.c [hunk 4]
```diff
@@ -812,18 +812,6 @@ PyArray_PyIntAsIntp_ErrMsg(PyObject *o, const char * msg)
      * Since it is the usual case, first check if o is an integer. This is
      * an exact check, since otherwise __index__ is used.
      */
-#if !defined(NPY_PY3K)
```
Comment (sethtroisi) [inline numpy/core/src/multiarray/conversion_utils.c]: I incorrectly thought this was non-trivial.

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 5]
```diff
@@ -3156,7 +3156,6 @@ ulong_arrtype_hash(PyObject *obj)
     return x;
 }
 
-#if (NPY_SIZEOF_INT != NPY_SIZEOF_LONG) || defined(NPY_PY3K)
```
Comment (sethtroisi) [inline numpy/core/src/multiarray/scalartypes.c.src]: I read these as `&&` which I thought was going to involve more cleanup.

File: numpy/core/src/multiarray/alloc.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/multiarray/alloc.c]: Yeah, I think that makes sense. Everything else looks good to me.

Comment (sethtroisi) [reply numpy/core/src/multiarray/alloc.c]: Inlined.

**Review by seberg — APPROVED** [review]
Thanks, just waiting for CI.

**Comment by seberg** [comment]
Thanks, tests all green (except flaky shippable), so putting it in.

**Pull request merged**